core/types/validator.go

```diff
@@ -27,63 +27,76 @@ func (val *Validator) Address() common.Address { return val.address }
 func (val *Validator) Deposit() uint64         { return val.deposit }
 func (val *Validator) Weight() *big.Int        { return val.weight }
 
-type ValidatorSet struct {
-	validators []*Validator
-	proposer   *Validator
-
-	//cache
-	membership map[common.Address]*Validator
+type ValidatorList interface {
+	UpdateWeights()
+	At(i int) *Validator
+	Get(addr common.Address) *Validator
+	Size() int
+	Proposer() *Validator
+	Contains(addr common.Address) bool
 }
 
-// @TODO (rgeraldes) - size needs to be > 0
-func NewValidatorSet(validators []*Validator) *ValidatorSet {
-	set := &ValidatorSet{
-		validators: validators,
-		membership: make(map[common.Address]*Validator, len(validators)),
+var ErrInvalidParams = errors.New("A validator set needs at least one validator")
+
+func NewValidatorList(validators []*Validator) (*validatorList, error) {
+	if len(validators) == 0 {
+		return nil, ErrInvalidParams
 	}
 
-	for _, validator := range validators {
-		set.membership[validator.address] = validator
+	set := &validatorList{
+		validators: validators,
+		proposer:   validators[0],
 	}
 
-	return set
+	return set, nil
+}
+
+type validatorList struct {
+	validators []*Validator
+	proposer   *Validator
 }
 
 // Update updates the weight and the proposer based on the set of validators
-func (set *ValidatorSet) UpdateWeight() {
-	pq := make(common.PriorityQueue, 0)
-	heap.Init(&pq)
+func (set *validatorList) UpdateWeights() {
+	proposer := set.validators[0]
 
 	for _, validator := range set.validators {
 		validator.weight = validator.weight.Add(validator.weight, big.NewInt(int64(validator.deposit)))
-		// @TODO (rgeraldes) - review types, possible overflow
-		heap.Push(&pq, &common.Item{Priority: int(validator.weight.Int64()), Value: validator})
+		if validator.weight.Cmp(proposer.weight) > 0 {
+			proposer = validator
+		}
 	}
-
-	item := heap.Pop(&pq).(*common.Item)
-	set.proposer = item.Value.(*Validator)
+	set.proposer = proposer
 
 	// decrement the validator weight since he has been selected
 	set.proposer.weight.Sub(set.proposer.weight, big.NewInt(int64(set.proposer.deposit)))
 }
 
-func (set *ValidatorSet) AtIndex(i int) *Validator {
+func (set *validatorList) At(i int) *Validator {
+	if i < 0 || i > len(set.validators) {
```

Comment:
actually `i > len(set.validators)` has to be `i >= len(set.validators)` right?

Reply:
better right a test for it :)